rt/transports/http-jetty/src/main/java/org/apache/cxf/transport/http_jetty/JettyHTTPServerEngine.java

```diff
@@ -937,25 +1035,81 @@ private ThreadPool getThreadPool() {
+        if (!smap.equals("/")) {
+            //add /* to enable wild match
+            smap = smap + "/*";
+        }
+        
+        if (contextName.endsWith("/*")) {
+            //This is how Jetty ServletContextHandler handle the contextName
+            //without suffix "/", the "*" suffix will be removed
+            contextName = contextName + "/";
+        }
 
         boolean found = false;
 
         if (server != null && server.isRunning()) {
-            for (Handler handler : contexts.getChildHandlersByClass(ContextHandler.class)) {
-                if (handler instanceof ContextHandler) {
-                    ContextHandler contextHandler = (ContextHandler) handler;
-                    Handler jh = contextHandler.getHandler();
-                    if (jh instanceof JettyHTTPHandler
+            for (Handler handler : contexts.getHandlers()) {
+                if (handler instanceof ServletContextHandler) {
+                    ServletContextHandler contextHandler = (ServletContextHandler) handler;
+                    ServletHandler servletHandler = contextHandler.getServletHandler();
+                    
+                    MatchedResource<MappedServlet> mappedServlet = servletHandler.getMatchedServlet(smap);
+                    if (mappedServlet == null) {
+                        continue;
+                    }
+                    ServletHolder servletHolder = mappedServlet.getResource().getServletHolder();
+                    Servlet servlet = null;
+                    
+                    try {
+                        servlet = servletHolder.getServlet();
+                    } catch (ServletException ex) {
+                        LOG.log(Level.WARNING, "REMOVE_HANDLER_FAILED_MSG", new Object[] {
+                                                                                       ex.getMessage()
+                        });
+                        continue;
+                    }
+                    if (servlet != null && servlet instanceof JettyHTTPHandler
                         && (contextName.equals(contextHandler.getContextPath())
                             || (StringUtils.isEmpty(contextName)
                                 && "/".equals(contextHandler.getContextPath())))
-                        && ((JettyHTTPHandler)jh).getName().equals(smap)) {
+                        && smap.startsWith(((JettyHTTPHandler)servlet).getName())) {
                         try {
-                            contexts.removeHandler(handler);
-                            handler.stop();
-                            handler.destroy();
+                            servletHolder.stop();
+                            contextHandler.getContext().destroy(servlet);
+                            //need to remove path from ServletHandler._servletMappings
+                            //and has to access the private field.
+                            
+                            Field privateField
```

Comment:
@ffang could we avoid using reflection? If we really really need that - please use `org.apache.cxf.common.util.ReflectionUtil` (but I would love to see the solution without reflection)

Reply:
We have to use ReflectionUtil here. A new commit addressed it